### PR TITLE
Sitemap crawling

### PR DESCRIPTION
**Pull request by alex-moreno**

### CODE REVIEW

File: src/ScraperBot/Command/CrawlSitesCommand.php [hunk 1]
```diff
@@ -68,7 +71,27 @@ protected function execute(InputInterface $input, OutputInterface $output) {
...
+
+        $timestamp = time();
+        $crawler->crawlSites($source, $client, $default_config, $timestamp);
+        $crawler->crawlSiteMaps($source, $client, $default_config, $timestamp, $source->getCurrentIndex());
+
+        $sitemapURLs = $crawler->getListPendingSitemaps(TRUE);
+        $sourceSitemap = new XmlSitemapSource($sitemapURLs);
+
+        // Crawl the sitemaps.
+        /* // NOT NEEDED A SECOND TIME -> $crawler->crawlSiteMaps($sourceSitemap, $client, $default_config, $timestamp, $source->getCurrentIndex()); */
```
Comment (alex-moreno) [inline src/ScraperBot/Command/CrawlSitesCommand.php:83]: remove

File: src/ScraperBot/Command/CrawlSitesCommand.php [hunk 2]
```diff
@@ -68,7 +71,27 @@ protected function execute(InputInterface $input, OutputInterface $output) {
...
+        $crawler->extractSitemaps($sourceSitemap, $client, $default_config, $timestamp, $source->getCurrentIndex());
+
+        $pendingURLs = $sqlStorage->getPendingURLs(TRUE);
+        $pendingSource = new SitesArraySource($pendingURLs);
+        $crawler->crawlSites($pendingSource, $client, $default_config, $timestamp);
+
+        // TODO:
+        //  DONE - 1. FETCH LIST OF SITES IN THE SITEMAP IN $sourceSitemap, and store in pendingURLs
+        //  DONE - 2. Crawl all sites in pendingURLs.
+        // 3. adapt sites.php to list all urls included added pendingURLs.
```
Comment (alex-moreno) [inline src/ScraperBot/Command/CrawlSitesCommand.php:93]: remove

File: src/ScraperBot/Crawler.php [hunk 3]
```diff
@@ -102,5 +114,140 @@ public function crawlSites(SourceInterface $source, Client $client, $default_con
...
+                    // We want to follow Sitemap: urls.
+                    if (strpos($line, 'Sitemap:') !== false) {
+                        $this->offIndex++;
+                        $newurl = trim(substr($line, strlen("Sitemap:"), strlen($line)));
+                        echo PHP_EOL . 'Storing sitemap: ' . $newurl . PHP_EOL;
+                        echo 'offindex: ' . $this->offIndex . PHP_EOL;
+                        // Store new links.
+                        $this->storage->addSitemapURL($newurl, $this->offIndex, $timestamp);
+                    }
+                }
```
Comment (alex-moreno) [inline src/ScraperBot/Crawler.php:178]: some DRY here, would be good to clean in the short future

File: src/ScraperBot/Source/CsvSource.php [hunk 4]
```diff
@@ -29,7 +28,21 @@ function($entry) {
...
+    public function addLink($url) {
+        $this->listOfSites[] = $url;
+    }
+
+    public function readLinks() {
+        return $this->listOfSites;
+    }
+
+
+    public function getCurrentIndex() {
```
Comment (alex-moreno) [inline src/ScraperBot/Source/CsvSource.php:45]: docblock

File: src/ScraperBot/Source/SitesArraySource.php [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+    public function __construct($sites) {
+        $this->sites = $sites;
+    }
+
+    public function getLinks()
+    {
+        return $this->sites;
+    }
+
+    public function addLink($url)
```
Comment (alex-moreno) [inline src/ScraperBot/Source/SitesArraySource.php:20]: docblock

File: src/ScraperBot/Storage/SqlLite3Storage.php [hunk 6]
```diff
@@ -115,14 +130,17 @@ public function getCrawlDiffs($timestamp1, $timestamp2, $tolerance = 1000) {
 
         $index2 = 1;
         while ($row2 = $results2->fetchArray()) {
-            $listofSites2[$index2] = $row2;
+            // url == $row2[2].
```
Comment (alex-moreno) [inline src/ScraperBot/Storage/SqlLite3Storage.php:133]: remove

File: src/ScraperBot/Storage/SqlLite3Storage.php [hunk 7]
```diff
@@ -149,4 +167,60 @@ public function getStatsByStatus($statusCode) {
         return $numRows;
     }
 
+    public function addSitemapURL($url, $index, $timestamp)
+    {
+        // TODO: Implement addTemporaryURL() method.
```
Comment (alex-moreno) [inline src/ScraperBot/Storage/SqlLite3Storage.php:172]: remove

File: src/ScraperBot/Storage/SqlLite3Storage.php [hunk 8]
```diff
@@ -149,4 +167,60 @@ public function getStatsByStatus($statusCode) {
...
+    }
+
+    public function dumpSitemaps() {
+        $this->pdo->query("delete from sitemapURLs ");
+    }
+
+
+    public function addPendingURL($url, $index, $timestamp)
+    {
+        // TODO: Implement addTemporaryURL() method.
```
Comment (alex-moreno) [inline src/ScraperBot/Storage/SqlLite3Storage.php:200]: remove

File: src/ScraperBot/Storage/SqlLite3Storage.php [hunk 9]
```diff
@@ -149,4 +167,60 @@ public function getStatsByStatus($statusCode) {
...
+
+        return $results;
+    }
+
+    public function dumpSitemaps() {
+        $this->pdo->query("delete from sitemapURLs ");
+    }
+
+
+    public function addPendingURL($url, $index, $timestamp)
```
Comment (alex-moreno) [inline src/ScraperBot/Storage/SqlLite3Storage.php:198]: docblocks

File: src/ScraperBot/Storage/SqlLite3Storage.php [hunk 10]
```diff
@@ -83,11 +132,15 @@ public function getStatusCodes() {
      * @param $timestamp
      * @return mixed
      */
-    public function getResultsbyTimestamp($timestamp) {
-        $queryString = sprintf("SELECT * FROM sites WHERE timestamp = '%s' GROUP BY `site_id` order by site_id", $timestamp);
+    public function getResultsbyTimestamp($timestamp, $onlyLatest = NULL) {
+        if ($onlyLatest != NULL) {
```
Comment (iainp999) [inline src/ScraperBot/Storage/SqlLite3Storage.php:136]: empty block?

File: html/sites.php [hunk 11]
```diff
@@ -2,8 +2,19 @@
...
+$resultsStorage = new \ScraperBot\Storage\SqlLite3Storage('../glitcherbot.sqlite3');
+
+$compare = [];
+if (isset($_GET['date1']) && $_GET['date2']) {
+    echo $_GET['date1'] . ' date2: ' . $_GET['date2'];
+    $compare['date1'] = $_GET['date1'];
+    $compare['date2'] = $_GET['date2'];
+}
+
+if (isset($_GET['latest'])) {
```
Comment (iainp999) [inline html/sites.php:14]: I'm not sure what the behaviour is for `onlyLatest`?

File: src/ScraperBot/Crawler.php [hunk 12]
```diff
@@ -102,5 +115,167 @@ public function crawlSites(SourceInterface $source, Client $client, $default_con
...
+    /**
+     * Crawl sites.
+     *
+     * @param $listOfSites
+     * @param $client
+     * @throws \GuzzleHttp\Exception\GuzzleException
+     */
+    public function crawlSiteMaps(SourceInterface $source, Client $client, $default_config = NULL, $timestamp = NULL, $offIndex = 0)
+    {
+        echo PHP_EOL . 'Sitemaps crawling>>>> ';
```
Comment (iainp999) [inline src/ScraperBot/Crawler.php:155]: If we start using echo, we're bypassing the symfony console.

Comment (alex-moreno) [reply src/ScraperBot/Crawler.php:155]: no, we should not. Thanks for raising it. I just need to check how to pass that class, maybe using DI?

File: html/sites.php (hunk 11 above)
Comment (alex-moreno) [reply html/sites.php:14]: show only latest results, but I'll get rid of that for now.

**Comment by alex-moreno** [comment]
merging this as more changes coming soon

**Pull request merged**